drivers/net/ethernet/intel/idpf/idpf_aux.c

```diff
@@ -0,0 +1,327 @@
+}
+
+/**
+ * idpf_aux_dispatch_event - Send event to auxiliary devices)
+ * @adapter: IDPF main private structure
+ * @code: IDC event code
+ * @dev_info: Auxiliary device info, NULL to dispatch to all devices
+ * @data: Event-specific data
+ *
+ * Returns number of auxiliary devices the event was dispatched to.
+ */
+int idpf_aux_dispatch_event(struct idpf_adapter *adapter,
+			    enum idpf_aux_idc_event_code code,
+			    struct idpf_aux_dev_info *dev_info,
+			    void *data)
+{
+	struct idpf_auxiliary_driver *aux_drv;
+	struct idpf_aux_idc_event event;
+	int i, devs = 0;
+
+	if (!adapter->adevs)
+		return devs;
+
+	aux_drv = idpf_eth_get_driver();
+
+	event.event_code = code;
+	event.event_data = data;
+	if (!dev_info) {
+		for (i = 0; i < idpf_get_default_vports(adapter); ++i) {
+			if (!adapter->adevs[i])
+				continue;
+
+			dev_info = &adapter->adevs[i]->aux_info;
+			aux_drv->event_handler(dev_info, &event);
+			devs++;
+		}
+	} else {
+		if (event.event_code == IDPF_AUX_IDC_EVENT_LINK_CHANGE) {
+			aux_drv->event_handler(dev_info, &event);
+			devs++;
+		}
+	}
+
+	return devs;
+}
+
+/**
+ * idpf_aux_recv_event - Recieve an event from auxiliary device
+ * @adapter: Auxiliary device info
+ * @event: IDC event
+ */
+static void idpf_aux_recv_event(struct idpf_aux_dev_info *dev_info,
+				struct idpf_aux_idc_event *event)
+{
+	struct idpf_auxiliary_device *aux_dev;
+	struct idpf_adapter *adapter;
+	struct device *dev;
+
+	aux_dev = container_of(dev_info, struct idpf_auxiliary_device,
+			       aux_info);
+	adapter = container_of(dev_info->aux_shared, struct idpf_adapter,
+			       aux_shared);
+	dev = idpf_aux_to_dev(aux_dev);
+
+	switch (event->event_code) {
+	case IDPF_AUX_IDC_EVENT_REQ_HARD_RESET:
+		if (!idpf_is_reset_in_prog(adapter)) {
+			set_bit(IDPF_HR_FUNC_RESET, adapter->flags);
+			queue_delayed_work(adapter->vc_event_wq,
+					   &adapter->vc_event_task,
+					   msecs_to_jiffies(10));
+		}
+		break;
+
+	default:
+		dev_err(dev, "Unknwon auxiliary event: %u\n",
+			event->event_code);
+		break;
+	}
+}
+
+/**
+ * idpf_aux_recv_virtchnl - Pass on a virtchnl msg on behalf of auxiliary dev
+ * @adapter: Auxiliary device info
+ * @params: Virtchnl2 parameters
+ */
+static ssize_t idpf_aux_recv_virtchnl(struct idpf_aux_dev_info *dev_info,
+				      struct idpf_vc_xn_params *params)
+{
+	struct idpf_adapter *adapter;
+
+	adapter = container_of(dev_info->aux_shared, struct idpf_adapter,
+			       aux_shared);
+	return idpf_vc_xn_exec(adapter, params);
+}
+
+/**
+ * idpf_aux_init_shared - Initialize the auxiliary shared info
+ * @adapter: IDPF main private structure
+ */
+void idpf_aux_init_shared(struct idpf_adapter *adapter)
+{
+	struct idpf_idc_ops *idc_ops;
+
+	adapter->aux_shared.hw_addr = adapter->hw.hw_addr;
+
+	idc_ops = &adapter->aux_shared.idc_ops;
+	idc_ops->event_send = idpf_aux_recv_event;
+	idc_ops->virtchnl_send = idpf_aux_recv_virtchnl;
+	idc_ops->intr_reg_init = adapter->dev_ops.reg_ops.intr_reg_init;
```

Comment:
Assuming it is now in idc_ops do we still need dev_ops assigment? Shouldn't it be dispatched during driver creation (different ops for PF eth driver, different for VF eth driver, and probably different for ixd devices). We shouldn't duplicate the places when this different is handled.

Type can be passed here with correct ops filled or sth like that.

Reply:
Ok, I was just wondering if we can move this initialization to the moment when eth aux device (VF or PF) is created.